packages/loader/container-loader/src/container.ts

```diff
@@ -162,6 +162,80 @@ export enum ConnectionState {
     Connected,
 }
 
+enum LifecycleState {
+    /**
+     * The container has been created
+     */
+    Created = "created",
+
+    /**
+     * The container is loading
+     */
+    Loading = "loading",
+
+    /**
+     * The container has been loaded
+     */
+    Loaded = "loaded",
+
+    /**
+     * The container has been marked for closure
+     */
+    Closing = "closing",
+
+    /**
+     * The container is in closed state
+     */
+    Closed = "closed",
+}
+
+class LifecycleStateHandler {
+    constructor(state: LifecycleState) {
+        this._state = state;
+    }
+
+    public get state(): LifecycleState {
+        return this._state;
+    }
+
+    /**
+     * Returns true if the container has been closed, otherwise false
+     */
+    public isClosed(): boolean {
+        return (this._state === LifecycleState.Closing || this._state === LifecycleState.Closed);
+    }
+
+    /**
+     * Returns true if the container has been loaded, otherwise false
+     */
+    public isLoaded(): boolean {
+        return (this._state !== LifecycleState.Created && this._state !== LifecycleState.Loading);
+    }
+
+    /**
+     * Transitions the container's lifecycle state
+     */
+    public changeState(newState: LifecycleState) {
+        // TODO: Figure our the possible state transitions...
```

Comment:
Looking at the code today, I think these are the requirements when going to each newState:

* created - not allowed
* loading - created
* loaded - not created
* closing - any
* closed - closing

Reply:
As for what to do if these requirements are violated - I'm unsure whether asserting is too harsh.  Other option is to pass a logger into the constructor here, and use `logIfFalse` instead of `assert`.

I guess looking at the list, I'm not actually worried about it... so let's stick with assert.

Reply:
I am wondering if, for both loading and loaded states, it is possible that the container has been marked for closure before the state transition happens?

I've tried the logger approach but if you feel we should be throwing the error instead then I can make the changes.